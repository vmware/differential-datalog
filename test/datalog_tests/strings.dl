ground relation Strings(s: string)

Strings("").
Strings("foo").
Strings("bar\n").
Strings("\t\tbar").
Strings("\"buzz").
Strings("\r\abuzz").
Strings("\\buzz").
Strings("\\buzz").

/* unicode */
Strings("barΔ").

/* raw strings */
Strings([||]).
Strings([|foo|]).
Strings([|foo
buzz
bar|]).

Strings([|
    foo
    buzz
    bar|]).

Strings([|barΔ
Π|]).

Strings([|barΔ
Π|] " quoted string").

// backslash does not work as an escape character in raw strings
Strings([|
    foo\r
    buzz
    bar|]).

/* string literal concatenation */
Strings("foo" "-" "bar").

Strings("foo\n" "bar").
Strings("foo\\""bar").

Strings("foo\\" [|bar|]).

function dummy(x: string): string = x ++ " dummy"

function strings(): () = {
    var str1 = "foo" ++ "bar";
    var str2 = "foo" "bar" ++ "buzz" ++ [|
raw string |]"quoted string";
    var str3 = $"str1: { str1 }";
    var str4 = $"str1: {str1}, str2: {str2}, str3: {str3}.";
    var str5 = $[|str1: {str1},
str2: {str2},
str3: {str3}.|];
    var str6 = $"{str1 ++ str2}";
    var str7 = $"{{var x = str1 ++ str2; x}}";
    var str8 = $"{{var x = str1 ++ str2; dummy(x)}}";
    ()
}

Strings("word1\
        \ word2").

Strings("line1\n\
        \line2").
