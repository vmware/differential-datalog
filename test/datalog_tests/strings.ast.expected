typedef Strings = Strings{s: string}
typedef serializable_t = ConsInt{x: int} | ConsBit{y: bit<32>} | ConsBool{z: bool} | Cons0{}
typedef set<'A>
function __builtin_2string (x: 'X): string
function dummy (x: string): string =
    (x ++ " dummy")
function serializable_t2string (x: serializable_t): string =
    match (x) {
        ConsInt{.x=var v} -> (("ConsInt(" ++ __builtin_2string(v)) ++ ")"),
        ConsBit{.y=var v} -> (("ConsBool(" ++ __builtin_2string(v)) ++ ")"),
        ConsBool{.z=true} -> "ConsBool(true)",
        ConsBool{.z=false} -> "ConsBool(false)",
        Cons0{} -> "Cons0"
    }
function strings (): () =
    (var str1 = ("foo" ++ "bar");
     (var str2 = (("foobar" ++ "buzz") ++ "\nraw string quoted string");
      (var str3 = (("str1: " ++ str1) ++ "");
       (var str4 = (((((("str1: " ++ str1) ++ ", str2: ") ++ str2) ++ ", str3: ") ++ str3) ++ ".");
        (var str5 = (((((("str1: " ++ str1) ++ ",\nstr2: ") ++ str2) ++ ",\nstr3: ") ++ str3) ++ ".");
         (var str6 = (("" ++ (str1 ++ str2)) ++ "");
          (var str7 = (("" ++ (var x = (str1 ++ str2);
                               x)) ++ "");
           (var str8 = (("" ++ (var x = (str1 ++ str2);
                                dummy(x))) ++ "");
            ()))))))))
function tostring1 (): () =
    ((var a: int) = 5;
     ((var b: bit<32>) = 32'd5;
      ((var c: bool) = true;
       ((var d: string) = "foo";
        (var err = (((((((("a=" ++ __builtin_2string(a)) ++ ", b=") ++ __builtin_2string(b)) ++ ", c=") ++ __builtin_2string(c)) ++ ", d=") ++ d) ++ "");
         ())))))
function tostring2 (): () =
    ((var a: int) = 5;
     ((var b: bit<32>) = 32'd5;
      ((var c: bool) = true;
       ((var d: string) = "foo";
        ((var e: serializable_t) = ConsInt{.x=0};
         (var err = (((((((((("a=" ++ __builtin_2string(a)) ++ ", b=") ++ __builtin_2string(b)) ++ ", c=") ++ __builtin_2string(c)) ++ ", d=") ++ d) ++ ", e:") ++ serializable_t2string(e)) ++ "");
          ()))))))
ground relation Strings [Strings]
Strings(.s="").
Strings(.s="foo").
Strings(.s="bar\n").
Strings(.s="\t\tbar").
Strings(.s="\"buzz").
Strings(.s="\r\n\abuzz").
Strings(.s="\\buzz").
Strings(.s="\\buzz").
Strings(.s="bar\916").
Strings(.s="").
Strings(.s="foo").
Strings(.s="foo\nbuzz\nbar").
Strings(.s="\n    foo\n    buzz\n    bar").
Strings(.s="bar\916\n\928").
Strings(.s="bar\916\n\928 quoted string").
Strings(.s="\n    foo\\r\\n\n    buzz\n    bar").
Strings(.s="foo-bar").
Strings(.s="foo\nbar").
Strings(.s="foo\\bar").
Strings(.s="foo\\bar").
Strings(.s="word1 word2").
Strings(.s="line1\nline2").
